File: src/libs/m3/vfs/MountTable.cc
/*
 * Copyright (C) 2016-2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/com/Marshalling.h>
#include <m3/com/GateStream.h>
#include <m3/session/M3FS.h>
#include <m3/vfs/MountTable.h>
#include <m3/vfs/VFS.h>
#include <m3/Exception.h>

namespace m3 {

static size_t charcount(const char *str, char c) {
    size_t cnt = 0;
    while(*str) {
        if(*str == c)
            cnt++;
        str++;
    }
    // if the path does not end with a slash, we have essentially one '/' more
    if(str[-1] != '/')
        cnt++;
    return cnt;
}

// TODO this is a very simple solution that expects "perfect" paths, i.e. with no "." or ".." and
// no duplicate slashes (at least not just one path):
static size_t is_in_mount(const String &mount, const char *in) {
    const char *p1 = mount.c_str();
    const char *p2 = in;
    while(*p2 && *p1 == *p2) {
        p1++;
        p2++;
    }
    while(*p1 == '/')
        p1++;
    if(*p1)
        return 0;
    while(*p2 == '/')
        p2++;
    return static_cast<size_t>(p2 - in);
}

MountTable::~MountTable() {
    for(size_t i = 0; i < MAX_MOUNTS; ++i)
        delete _mounts[i];
}

void MountTable::add(const char *path, Reference<FileSystem> fs) {
    if(_count == MAX_MOUNTS)
        throw MessageException("No free slot in mount table", Errors::NO_SPACE);

    size_t compcount = charcount(path, '/');
    size_t i = 0;
    for(; i < _count; ++i) {
        // mounts are always tightly packed
        assert(_mounts[i]);

        if(strcmp(_mounts[i]->path().c_str(), path) == 0)
            VTHROW(Errors::EXISTS, "Mountpoint " << path << " already exists");

        // sort them by the number of slashes
        size_t cnt = charcount(_mounts[i]->path().c_str(), '/');
        if(compcount > cnt)
            break;
    }
    assert(i < MAX_MOUNTS);

    // move following items forward
    if(_count > 0) {
        for(size_t j = _count; j > i; --j)
            _mounts[j] = _mounts[j - 1];
    }
    _mounts[i] = new MountPoint(path, fs);
    // ensure that we don't reuse ids, even if this filesystem was added after unserialization
    _next_id = Math::max(fs->id() + 1, _next_id);
    _count++;
}

Reference<FileSystem> MountTable::resolve(const char **path, char *buffer, size_t bufsize) {
    auto res = try_resolve(path, buffer, bufsize);
    if(res)
        return res;
    VTHROW(Errors::NO_SUCH_FILE, "Unable to resolve path '" << *path << "'");
}

Reference<FileSystem> MountTable::try_resolve(const char **path, char *buffer, size_t bufsize) noexcept {
    if(**path != '/') {
        OStringStream os(buffer, bufsize);
        const char *cwd = VFS::cwd();
        os << cwd;
        if(strcmp(cwd, "/") != 0)
            os << "/";
        os << *path;
        *path = buffer;
    }

    for(size_t i = 0; i < _count; ++i) {
        size_t pos = is_in_mount(_mounts[i]->path(), *path);
        if(pos != 0) {
            *path = *path + pos;
            return _mounts[i]->fs();
        }
    }
    return Reference<FileSystem>();
}

Reference<FileSystem> MountTable::get_by_id(size_t id) noexcept {
    for(size_t i = 0; i < _count; ++i) {
        if(_mounts[i]->fs()->id() == id)
            return _mounts[i]->fs();
    }
    return Reference<FileSystem>();
}

const char *MountTable::path_of_id(size_t id) noexcept {
    for(size_t i = 0; i < _count; ++i) {
        if(_mounts[i]->fs()->id() == id)
            return _mounts[i]->path().c_str();
    }
    return nullptr;
}

size_t MountTable::indexof_mount(const char *path) {
    for(size_t i = 0; i < _count; ++i) {
        if(strcmp(_mounts[i]->path().c_str(), path) == 0)
            return i;
    }
    return MAX_MOUNTS;
}

void MountTable::remove(const char *path) {
    size_t idx = indexof_mount(path);
    if(idx != MAX_MOUNTS)
        do_remove(idx);
}

void MountTable::remove_all() noexcept {
    while(_count > 0)
        do_remove(0);
}

void MountTable::do_remove(size_t i) {
    assert(_mounts[i] != nullptr);
    assert(_count > 0);
    delete _mounts[i];
    // move following items backwards
    for(; i < _count - 1; ++i)
        _mounts[i] = _mounts[i + 1];
    _count--;
    _mounts[_count] = nullptr;
}

size_t MountTable::serialize(ChildActivity &act, void *buffer, size_t size) const {
    Marshaller m(static_cast<unsigned char*>(buffer), size);

    m << act._mounts.size();
    for(auto mapping = act._mounts.begin(); mapping != act._mounts.end(); ++mapping) {
        auto mount = Activity::own().mounts()->get(mapping->second.c_str());
        auto type = mount->type();
        m << mapping->first << type;
        switch(type) {
            case 'M':
                mount->serialize(m);
                break;
        }
    }
    return m.total();
}

void MountTable::delegate(ChildActivity &act) const {
    for(auto mapping = act._mounts.begin(); mapping != act._mounts.end(); ++mapping) {
        auto mount = Activity::own().mounts()->get(mapping->second.c_str());
        char type = mount->type();
        switch(type) {
            case 'M':
                mount->delegate(act);
                break;
        }
    }
}

MountTable *MountTable::unserialize(const void *buffer, size_t size) {
    MountTable *ms = new MountTable();
    Unmarshaller um(static_cast<const unsigned char*>(buffer), size);
    size_t count;
    um >> count;
    while(count-- > 0) {
        char type;
        String path;
        um >> path >> type;
        switch(type) {
            case 'M':
                ms->add(path.c_str(), Reference<FileSystem>(M3FS::unserialize(um)));
                break;
        }
    }
    return ms;
}

void MountTable::print(OStream &os) const noexcept {
    os << "Mounts:\n";
    for(size_t i = 0; i < _count; ++i)
        os << "  " << _mounts[i]->path() << ": " << _mounts[i]->fs()->type() << "\n";
}

}

File: src/libs/rust/m3/src/vfs/mounttable.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use core::fmt;

use crate::borrow::StringRef;
use crate::cap::Selector;
use crate::cell::RefCell;
use crate::col::{String, ToString, Vec};
use crate::errors::{Code, Error};
use crate::rc::Rc;
use crate::serialize::Source;
use crate::session::M3FS;
use crate::tiles::{ChildActivity, StateSerializer};
use crate::vfs::{FileSystem, VFS};

/// A reference to a file system.
pub type FSHandle = Rc<RefCell<dyn FileSystem>>;

/// Represents a mount point
pub struct MountPoint {
    path: String,
    fs: FSHandle,
}

impl MountPoint {
    /// Creates a new mount point for given path and file system.
    pub fn new(path: &str, fs: FSHandle) -> MountPoint {
        MountPoint {
            path: path.to_string(),
            fs,
        }
    }
}

/// The table of mount points.
#[derive(Default)]
pub struct MountTable {
    mounts: Vec<MountPoint>,
    next_id: usize,
}

impl MountTable {
    /// Allocates a new id for the next file system
    pub fn alloc_id(&mut self) -> usize {
        let res = self.next_id;
        self.next_id += 1;
        res
    }

    /// Adds a new mount point at given path and given file system to the table.
    pub fn add(&mut self, path: &str, fs: FSHandle) -> Result<(), Error> {
        if self.path_to_idx(path).is_some() {
            return Err(Error::new(Code::Exists));
        }

        let pos = self.insert_pos(path);
        // ensure that we don't reuse ids, even if this filesystem was added after unserialization
        self.next_id = self.next_id.max(fs.borrow().id());
        self.mounts.insert(pos, MountPoint::new(path, fs));
        Ok(())
    }

    /// Returns the file system mounted exactly at the given path.
    pub fn get_by_path(&self, path: &str) -> Option<FSHandle> {
        self.path_to_idx(path).map(|i| self.mounts[i].fs.clone())
    }

    /// Returns the mount point with id `mid`.
    pub fn get_by_id(&self, mid: usize) -> Option<FSHandle> {
        self.mounts
            .iter()
            .find(|mp| mp.fs.borrow().id() == mid)
            .map(|mp| mp.fs.clone())
    }

    /// Returns the mount path of the mount with given id
    pub fn path_of_id(&self, mid: usize) -> Option<&String> {
        self.mounts
            .iter()
            .find(|mp| mp.fs.borrow().id() == mid)
            .map(|mp| &mp.path)
    }

    /// Resolves the given path to the file system image and the offset of the mount point within
    /// the path. The given path is turned into an absolute path in case it's relative. The returned
    /// offset refers to the absolute path.
    pub fn resolve(&self, path: &mut StringRef<'_>) -> Result<(FSHandle, usize), Error> {
        if !path.starts_with('/') {
            path.set(VFS::cwd() + "/" + &*path);
        }

        for m in &self.mounts {
            if path.starts_with(m.path.as_str()) {
                return Ok((m.fs.clone(), m.path.len()));
            }
        }
        Err(Error::new(Code::NoSuchFile))
    }

    /// Removes the mount point at `path` from the table.
    pub fn remove(&mut self, path: &str) -> Result<(), Error> {
        match self.path_to_idx(path) {
            Some(i) => {
                self.mounts.remove(i);
                Ok(())
            },
            None => Err(Error::new(Code::NoSuchFile)),
        }
    }

    pub(crate) fn delegate(&self, act: &ChildActivity) -> Result<Selector, Error> {
        let mut max_sel = 0;
        let mounts = act.mounts().clone();
        for (_cpath, ppath) in &mounts {
            if let Some(fs) = self.get_by_path(ppath) {
                let sel = fs.borrow().delegate(act)?;
                max_sel = sel.max(max_sel);
            }
        }
        Ok(max_sel)
    }

    pub(crate) fn serialize(&self, map: &[(String, String)], s: &mut StateSerializer<'_>) {
        s.push_word(map.len() as u64);

        for (cpath, ppath) in map {
            if let Some(fs) = self.get_by_path(ppath) {
                let fs = fs.borrow();
                let fs_type = fs.fs_type();
                s.push_str(cpath);
                s.push_word(fs_type as u64);
                fs.serialize(s);
            }
        }
    }

    pub(crate) fn unserialize(s: &mut Source<'_>) -> MountTable {
        let mut mt = MountTable::default();

        let count = s.pop().unwrap();
        for _ in 0..count {
            let path: String = s.pop().unwrap();
            let fs_type: u8 = s.pop().unwrap();
            mt.add(&path, match fs_type {
                b'M' => M3FS::unserialize(s),
                _ => panic!("Unexpected fs type {}", fs_type),
            })
            .unwrap();
        }

        mt
    }

    fn path_to_idx(&self, path: &str) -> Option<usize> {
        // TODO support imperfect paths
        assert!(path.starts_with('/'));
        assert!(path.ends_with('/'));
        assert!(!path.contains(".."));

        for (i, m) in self.mounts.iter().enumerate() {
            if m.path == path {
                return Some(i);
            }
        }
        None
    }

    fn insert_pos(&self, path: &str) -> usize {
        let comp_count = Self::path_comps(path);
        for (i, m) in self.mounts.iter().enumerate() {
            let cnt = Self::path_comps(m.path.as_str());
            if comp_count > cnt {
                return i;
            }
        }
        self.mounts.len()
    }

    fn path_comps(path: &str) -> usize {
        let mut comp_count = path.chars().filter(|&c| c == '/').count();
        if !path.ends_with('/') {
            comp_count += 1;
        }
        comp_count
    }
}

impl fmt::Debug for MountTable {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "MountTable[")?;
        for m in self.mounts.iter() {
            writeln!(f, "  {} -> {:?}", m.path, m.fs.borrow())?;
        }
        write!(f, "]")
    }
}
